torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,917 @@
+    }
+
+    for (auto neighbor_neighbor : neighbors[i].group->getNeighbors()) {
+      // Don't check self
+      if (neighbor_neighbor == neighbors[i].group) {
+        continue;
+      }
+      if (neighbor_neighbor->merged) {
+        // check neighbor_neighbor level
+        if (std::abs(neighbor_neighbor->level - level) <= 1) {
+          can_merge[i] = false;
+        }
+        if (std::abs(neighbor_neighbor->level - neighbors[i].group->level) <=
+            1) {
+          can_merge[i] = false;
+        }
+
+        // check neighbor_neighber->merged->level
+        if (std::abs(neighbor_neighbor->merge_with->level - level) <= 1) {
+          can_merge[i] = false;
+        }
+        if (std::abs(
+                neighbor_neighbor->merge_with->level -
+                neighbors[i].group->level) <= 1) {
+          can_merge[i] = false;
+        }
+      }
+    }
+  }
+
+  std::vector<NeighborGroup> merge_candidates;
+  for (size_t i = 0; i < neighbors.size(); i++) {
+    if (can_merge[i]) {
+      merge_candidates.push_back(neighbors[i]);
+    }
+  }
+  return merge_candidates;
+}
+
+void SegmentedGroup::clearTraversalInfo() {
+  level = -1;
+  visited = false;
+  merge_with = nullptr;
+  merged = false;
+}
+
+std::vector<Val*> SegmentedGroup::edgesToVals(
+    const std::vector<SegmentedEdge*>& se_v) {
+  std::vector<Val*> ret_v;
+  ret_v.reserve(se_v.size());
+
+  std::transform(
+      se_v.cbegin(),
+      se_v.cend(),
+      std::back_inserter(ret_v),
+      [](SegmentedEdge* se) { return se->val; });
+  return ret_v;
+}
+
+void SegmentedGroup::print() {
+  std::cout << "g{"
+            << "(" << toString(heuristic_) << ")\n";
+  for (size_t i = 0; i < exprs_.size(); i++) {
+    exprs_[i]->print();
+    if (i + 1 != exprs_.size())
+      std::cout << ", ";
+  }
+  std::cout << "}\n\n";
+}
+
+template <typename PREDICATE>
+void insertUniquePredicated(
+    std::vector<Val*>& v,
+    const std::vector<SegmentedEdge*>& e,
+    PREDICATE pred) {
+  std::unordered_set<Val*> to_add;
+  std::transform(
+      e.cbegin(),
+      e.cend(),
+      std::inserter(to_add, to_add.end()),
+      [](SegmentedEdge* se) { return se->val; });
+  std::copy_if(
+      to_add.begin(), to_add.end(), std::back_inserter(v), [pred](Val* val) {
+        return pred(val);
+      });
+}
+
+void SegmentedGroup::finalize() {
+  // Move all the edgees to group input/output
+  // Inputs
+  insertUniquePredicated(
+      input_vals, producer_edges, [](Val* v) { return !v->isFusionInput(); });
+
+  // Outputs
+  insertUniquePredicated(
+      output_vals, consumer_edges, [](Val* v) { return !v->isFusionOutput(); });
+}
+
+std::ostream& operator<<(std::ostream& os, const SegmentedGroup* group) {
+  os << "g{";
+  for (size_t i = 0; i < group->exprs().size(); i++) {
+    os << group->exprs()[i]->name();
+    if (i + 1 != group->exprs().size())
+      os << ", ";
+  }
+  os << "}\n";
+  return os;
+}
+
+void SegmentedEdge::print() {
+  std::cout << "e{ \n  ";
```

Comment:
same here - consolidate formatting in one place?

Reply:
Updated, Thanks!